src/ameba/rule/typing/method_param_type_restriction.cr

```diff
@@ -0,0 +1,102 @@
+module Ameba::Rule::Typing
+  # A rule that enforces method parameters have type restrictions, with optional enforcement of block parameters
+  #
+  # For example, these are considered valid:
+  #
+  # ```
+  # def listen(a : String, b : String) : String
+  #   a + b
+  # end
+  #
+  # def hello(name : String?)
+  #   puts "Hello, " + (name || "there") + "!"
+  # end
+  # ```
+  #
+  # And these are considered invalid:
+  #
+  # ```
+  # def listen(a, b) : String
+  #   a + b
+  # end
+  #
+  # def hello(name)
+  #   puts "Hello, " + (name || "there") + "!"
+  # end
+  # ```
+  #
+  # When the config options `PrivateMethods` and `ProtectedMethods`
+  # are true, this rule is also applied to private and protected methods, respectively.
+  #
+  # The `BlockParam` configuration option will extend this to block params, where this is valid:
+  #
+  # ```
+  # def exec(&block : String -> String)
+  #   yield "cmd"
+  # end
+  # ```
+  #
+  # And these are invalid:
+  #
+  # ```
+  # def exec(&)
+  # end
+  #
+  # def exec(&block)
+  # end
+  # ```
+  #
+  # The config option `Undocumented` controls whether this rule applies to undocumented methods and methods with a `:nodoc:` directive.
+  #
+  # The config option `DefaultValue` controls whether this rule applies to parameters that have a default value.
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/MethodParamTypeRestriction:
+  #   Enabled: false
+  #   DefaultValue: false
+  #   Undocumented: false
+  #   PrivateMethods: false
+  #   ProtectedMethods: false
+  #   BlockParam: false
+  # ```
+  class MethodParamTypeRestriction < Base
+    properties do
+      description "Recommends that method parameters have type restrictions"
+      enabled false
+      default_value true
+      undocumented false
```

Comment:
This option doesn't make sense to me. What relation does it have to this rule?

Reply:
When false, only documented methods will have this rule enforced (`:nodoc:` is considered undocumented). When true, all methods whether they're documented or not will have this rule enforced.

Reply:
I understand how it works, what I was asking was what's the rationale behind it and why do you think it's related to this rule in general.

Reply:
My thought process was for enabling this rule on the Crystal stdlib, where we would only want this to be enforced for the public API. Meaning this rule wouldn't / shouldn't apply to any methods that are undocumented / have nodoc directive.

Reply:
I'm not convinced tbh. I consider methods with `public` visibility modifier - not the documentation - as a part of the public API.

I'd leave it out for now, we can always add it l8r on.